gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb

```diff
@@ -0,0 +1,1078 @@
+      "outputs": [],
+      "source": [
+        "import json\n",
+        "from typing import List, Literal\n",
+        "import uuid\n",
+        "\n",
+        "from google.cloud import storage\n",
+        "from langchain_core.documents import Document\n",
+        "from langchain_core.messages import BaseMessage, HumanMessage\n",
+        "from langchain_core.tools import tool\n",
+        "from langchain_google_cloud_sql_pg import PostgresEngine, PostgresVectorStore\n",
+        "from langchain_google_vertexai import ChatVertexAI, VertexAIEmbeddings\n",
+        "from langgraph.graph import END, MessageGraph\n",
+        "from langgraph.prebuilt import ToolNode\n",
+        "import vertexai\n",
+        "from vertexai.preview import reasoning_engines"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "yPKXjZrFZuUZ"
+      },
+      "source": [
+        "### Authenticate to Google Cloud\n",
+        "\n",
+        "Authenticate to Google Cloud as the IAM user logged into this notebook in order to access your Google Cloud Project."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 2,
+      "metadata": {
+        "id": "NyKGtVQjgx13"
+      },
+      "outputs": [],
+      "source": [
+        "import sys\n",
+        "\n",
+        "if \"google.colab\" in sys.modules:\n",
+        "    from google.colab import auth\n",
+        "\n",
+        "    auth.authenticate_user()"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "9aGBuLA7aQ6O"
+      },
+      "source": [
+        "### Define project information\n",
+        "\n",
+        "Initialize `gcloud` with your Project ID and resource location. At this time, only `us-central1` is supported."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "DF4l8DTdWgPY"
+      },
+      "source": [
+        "### Set Google Cloud project information and initialize Vertex AI SDK\n",
+        "\n",
+        "To get started using Vertex AI, you must have an existing Google Cloud project and [enable the Vertex AI API](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com).\n",
+        "\n",
+        "Learn more about [setting up a project and a development environment](https://cloud.google.com/vertex-ai/docs/start/cloud-environment)."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "Nqwi-5ufWp_B"
+      },
+      "outputs": [],
+      "source": [
+        "PROJECT_ID = \"reasoning-engine-test-1\"  # @param {type:\"string\"}\n",
+        "LOCATION = \"us-central1\"  # @param {type:\"string\"}\n",
+        "STAGING_BUCKET = \"gs://reasoning-engine-test-1-bucket\"  # @param {type:\"string\"}\n",
+        "\n",
+        "vertexai.init(project=PROJECT_ID, location=LOCATION, staging_bucket=STAGING_BUCKET)\n",
+        "!gcloud config set project {PROJECT_ID}"
```

Comment:
Change the PROJECT_ID and STAGING_BUCKET to the values in the notebook template (https://github.com/GoogleCloudPlatform/generative-ai/blob/main/notebook_template.ipynb), such as: `[your-project-id]` and `[your-bucket-name]`.

Reply:
Changed.